subprojects/core/src/main/java/org/gradle/cache/internal/SimpleStateCache.java

```diff
@@ -59,14 +59,51 @@ public void run() {
         });
     }
 
-    public void update(final UpdateAction<T> updateAction) {
-        fileAccess.updateFile(new Runnable() {
+    public T update(final UpdateAction<T> updateAction) {
+        class Updater implements Runnable {
+            private final UpdateAction<T> updateAction;
+
+            private T result;
+
+            private Updater(UpdateAction<T> updateAction) {
+                this.updateAction = updateAction;
+            }
+
             public void run() {
                 T oldValue = deserialize();
-                T newValue = updateAction.update(oldValue);
-                serialize(newValue);
+                result = updateAction.update(oldValue);
+                serialize(result);
             }
-        });
+        }
+
+        Updater updater = new Updater(updateAction);
+        fileAccess.updateFile(updater);
+        return updater.result;
+    }
+
+    @Override
+    public T maybeUpdate(UpdateAction<T> updateAction) {
+        class MaybeUpdater implements Runnable {
+            private final UpdateAction<T> updateAction;
+
+            private T result;
+
+            private MaybeUpdater(UpdateAction<T> updateAction) {
+                this.updateAction = updateAction;
+            }
+
+            public void run() {
+                T oldValue = deserialize();
+                result = updateAction.update(oldValue);
+                if (oldValue != result) { // note: intentional pointer comparison
```

Comment:
Hey @ldaley, why is it intensional?

Reply:
I think this is a relic of first trying to jam the “maybe” semantics into the existing `update()` method. We could change this. 

Is it causing you problems?

Reply:
I've seen that a CI sometimes detects changes to `user-id.txt` when it shouldn't since it's running in the same container each time.

And a brief look at the code lead me to this line. I think it might be because strings will be the same but different references (for example one string is in String Pool and the other was created via `new`). In this case `Gradle` re-write `user-id.txt` and that way changes timestamp of the file which makes CI think it was changed.

Not a big issue at all since it rarely happens but I was wondered why this line is intensional instead something like `Objects.equal(oldValue, result)`

Reply:
@fkorotkov I'm going to change this for Gradle 4.6 to use logical equality.